components/common/src/templating/mod.rs

```diff
@@ -91,6 +94,50 @@ impl TemplateRenderer {
             .render(template, &raw)
             .map_err(|e| Error::TemplateRenderError(format!("{}", e)))
     }
+
+    // This method is only implemented so we can intercept the call to Handlebars and display
+    // a deprecation message to users. More information here https://github.com/habitat-sh/habitat/issues/6323.
+    // When Handlebars is upgraded this can be safely removed.
+    pub fn register_template_file<P>(&mut self,
+                                     name: &str,
+                                     path: P)
+                                     -> result::Result<(), TemplateFileError>
+        where P: AsRef<std::path::Path>
+    {
+        let path = path.as_ref();
+
+        lazy_static! {
+            static ref RE: Regex =
+                Regex::new(r"(\{\{[^}]+[^.])(\[[^}]*\}\})").expect("Failed to compile template \
+                                                                    deprecation regex");
+        }
+
+        let template_string =
+            std::fs::read_to_string(path).map_err(|e| {
+                                             TemplateFileError::IOError(e, name.to_owned())
+                                         })?;
+
+        template_string.split("\n")
+            .enumerate()
+            .filter(|(_i, v)| RE.is_match(&v))
+            .for_each(|(i, v)| {
+                println!("\n\n***************************************************\n\
+                          Deprecated object access syntax in handlebars template\n\n\
+                          TEMPLATE: {}\n\
+                          LINE: {}: '{}'\n\n\
+                          Use 'object.[index]' syntax instead of 'object[index]'\n\
+                          See https://github.com/habitat-sh/habitat/issues/6323 for more information\n\
+                          *******************************************************\n\n",
+                         path.display(), i, v)
+            });
+
+        // Replace the deprecated syntax with the good syntax. This should make it easier to upgrade
+        // the handlebars crate without breaking folks.
+        let updated_template_string = RE.replace_all(&template_string, "$1.$2");
```

Comment:
If we change things on the fly, we should output the changes we're making as well as provide a way to disable this behavior in case our changes are wrong.

The current code won't handle something like `{{ foo[bar][baz] }}`. I think we may be better off not using regular expressions and preferring a more direct approach.

Reply:
I'm not sure what you mean by "a more direct approach" but this does make me lean towards us not updating the templates on the fly. It seems like a lot of trouble for something that will shortly be deleted.

Reply:
It also adds another vector for bugs to creep in.

Reply:
> I'm not sure what you mean by "a more direct approach"

Using methods on `str`/`String` to find `[` not preceded by `.` in a handlebars context and replace them with `.[`. It's just a suggestion. If you think regular expressions would work better (or a hybrid approach), you can pursue that. I'm just pointing out an instance that the current code wouldn't handle. I typically find it easier to follow code that doesn't use regular expressions, so that's why I suggested it.

As for not updating the templates on the fly, I could go either way. The benefit is that we can delay (possibly indefinitely) breaking old templates without delaying updating our `handlebars` crate. But you're right that it does add some complexity. I don't know if we have a commitment to the timeline for when we'd flip the switch from deprecation to unsupported, but if it has significant impact on users, I think it's worth a little complexity.